### PR TITLE
Removing Reakit `Composite` implementation

### PR DESCRIPTION
**Pull request by andrewhayward**
## What?

Removes the 'unstable' Reakit implementation of the `Composite` component group.

## Why?

We need to remove Reakit from the codebase (#53278).

## How?

* Points default export to `./legacy` implementation
* Changes 'legacy' Storybook title to "Composite"
* Removes deprecation messages from _legacy_ (to be put back when _current_ is available)
* Removes _unstable_ Reakit implementation

## Testing Instructions

Open Storybook, and confirm that there are still two `Composite` entries ("Composite" and "Composite (V2)").

Storybook sidebar items, with "Composite (V2)" and "Composite" highlighted.

> [!NOTE]
> There are no consumers of the Reakit implementation left in Gutenberg, so we have to trust that test coverage is sufficient that any switch is transparent for consumers.

### CODE REVIEW

File: packages/components/src/composite/index.ts [hunk 1]
```diff
@@ -1,4 +1,8 @@
-// Until we migrate away from Reakit, the 'unstable'
-// implementation remains the default.
+// Originally this pointed at a Reakit implementation of
+// `Composite`, but we are removing Reakit entirely from the
+// codebase. Until the new 'current' Ariakit implementation
+// is unlocked and no longer private, we will continue to
+// support the Reakit API through the 'legacy' version,
+// which uses Ariakit under the hood.
```
Comment (ciampo) [inline packages/components/src/composite/index.ts]: We will continue to support this older implementation also after the new `Ariakit` implementation is unlocked, at least for a while

```suggestion
// codebase. We will continue to support this older implementation
// through a legacy version which uses Ariakit under the hood.
```

File: packages/components/src/composite/legacy/index.tsx [hunk 2]
```diff
@@ -99,14 +98,6 @@ type CompositeComponentProps = CompositeState &
 		| ComponentProps< typeof Current.CompositeRow >
 	);
 
-function showDeprecationMessage( previous?: string, next?: string ) {
```
Comment (ciampo) [inline packages/components/src/composite/legacy/index.tsx:102]: We should update unit tests to reflect the lack of a deprecation warning

**Review by ciampo — APPROVED** [review]
Left a couple of comments, but overall the changes look good.

Should we also remove `reakit` in this PR, or do you prefer to do it in a follow-up?

(also cc @mirka for an extra pair of eyes)

**Comment by andrewhayward** [comment]
> Should we also remove `reakit` in this PR, or do you prefer to do it in a follow-up?

Personally I'd rather do it in a follow-up PR, if only to keep this one focused on `Composite`, but that's not a strongly-held opinion.

File: packages/components/CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +28,10 @@
 
 -   `CheckboxControl`: Remove ability for label prop to be false ([#58339](https://github.com/WordPress/gutenberg/pull/58339)).
 
+### Code Quality
```
Comment (tyxla) [inline packages/components/CHANGELOG.md]: Dependency removals are usually "Internal" IIRC.

```suggestion
### Internal
```

File: packages/components/CHANGELOG.md [hunk 4]
```diff
@@ -28,6 +28,10 @@
```
Comment (tyxla) [inline packages/components/CHANGELOG.md:28]: Also unrelated, but just spotted that we have 2 "Enhancements" sections.

**Review by tyxla — APPROVED** [review]
Nice 👍

**Comment by ciampo** [comment]
Adding the `Needs dev note` label.

Assuming that changes up to reakit removal will be part of the WordPress 6.5 release, and that the promotion of the new `Composite` component from private to public API will be part of the WordPress 6.6 release, we will need to separate dev notes:

- one for the 6.5 release, highlighting that we removed `reakit` because of its incompatibility with newer versions of react and wrote a back-compat layer in `ariakit` which should cover standard use cases of the component (in theory no action required from consumers of the package). We could also anticipate that we're looking into exposing a new, stable version of the component written natively in `ariakit`.
- one for the 6.6 release, where we'll announce the new, stable, `ariakit`-based `Composite` component, and the related deprecation of the `__unstableComposite` family of components.

cc @mirka and @bph

**Comment by andrewhayward** [comment]
## Dev note

### Changes to the underlying _Composite_ component implementation

WordPress 6.5 no longer includes the [Reakit](https://reakit.io) package, as it does not support React beyond v16. The `__unstable*` [Composite component](https://github.com/WordPress/gutenberg/tree/5319e96303b515f98feb7925ae5a2e94cd46c4c8/packages/components/src/composite/legacy) was built on this library, so has been internally refactored.

#### What does this mean for consumers?

The primary API has not changed, and can still be imported with no changes. For typical usage patterns, there should be no differences, and in most cases no action is required from consumers of the package.

However, composite state props _must_ now be generated by the `useCompositeState` hook; they can no longer be provided by independent state logic. [Composite state arguments](https://wordpress.github.io/gutenberg/?path=/docs/components-composite--docs#tabbutton-prop-table-div-usecompositestate) have not changed though, and will continue to work as before.

```jsx
import {
  __unstableComposite: Composite,
  __unstableCompositeGroup: CompositeGroup,
  __unstableCompositeItem: CompositeItem,
  __unstableUseCompositeState: useCompositeState
} from '@wordpress/components';

const state = useCompositeState({ ... });

// ✅ This will continue to work
...( <Composite { ...state } /> );

// ⛔️ This will no longer work
...( <Composite { ...state } currentId={ ... } /> );
```

Consumers can continue to either spread the state, or pass as a single `state` prop...

```jsx
// ✅ Used by spreading the state
...(
  <Composite { ...state }>
    <CompositeGroup { ...state }>
      <CompositeItem { ...state }>
        { ... }
      </CompositeItem>
    </CompositeGroup>
  </Composite>
);

// ✅ Or with a single `state` prop
...(
  <Composite state={ state }>
    <CompositeGroup state={ state }>
      <CompositeItem state={ state }>
        { ... }
      </CompositeItem>
    </CompositeGroup>
  </Composite>
);
```

Because the shape of the returned composite state has changed, consumers can also now no longer destructure specific state props from `useCompositeState`.

```jsx
// ✅ This will continue to work
const state = useCompositeState({ ... });

// ⛔️ This will no longer work
const { groups, items } = useCompositeState({ ... });
```

#### What's next?

We anticipate a new stable _Composite_ component to be released in WordPress 6.6, along with the deprecation of this unstable version.

**Pull request merged**